src/store_client.cc

```diff
@@ -509,30 +551,43 @@ store_client::readBody(const char *, ssize_t len)
     if (len < 0)
         return fail();
 
-    const auto rep = entry->mem_obj ? &entry->mem().baseReply() : nullptr;
-    if (copyInto.offset == 0 && len > 0 && rep && rep->sline.status() == Http::scNone) {
-        /* Our structure ! */
-        if (!entry->mem_obj->adjustableBaseReply().parseCharBuf(copyInto.data, headersEnd(copyInto.data, len))) {
-            debugs(90, DBG_CRITICAL, "Could not parse headers from on disk object");
-        } else {
-            parsed_header = 1;
-        }
+    const auto expectingHeader = expectingHttpHeader();
+    auto doneParsingHeader = false;
+    if (expectingHeader) {
+        doneParsingHeader = parseHttpHeader(len);
+        if (!object_ok)
+            return;
     }
 
+    const auto rep = entry->mem_obj ? &entry->mem().baseReply() : nullptr;
     if (len > 0 && rep && entry->mem_obj->inmem_lo == 0 && entry->objectLen() <= (int64_t)Config.Store.maxInMemObjSize && Config.onoff.memory_cache_disk) {
         storeGetMemSpace(len);
         // The above may start to free our object so we need to check again
         if (entry->mem_obj->inmem_lo == 0) {
             /* Copy read data back into memory.
              * copyInto.offset includes headers, which is what mem cache needs
              */
-            int64_t mem_offset = entry->mem_obj->endOffset();
-            if ((copyInto.offset == mem_offset) || (parsed_header && mem_offset == rep->hdr_sz)) {
+            const auto memOffset = entry->mem_obj->endOffset();
+            if (doneParsingHeader) {
+                assert(replyHeaderBytes);
+                entry->mem_obj->write(StoreIOBuffer(replyHeaderBytes, 0));
+            } else if (copyInto.offset == memOffset && memOffset > 0) {
+                assert(!expectingHeader);
                 entry->mem_obj->write(StoreIOBuffer(len, copyInto.offset, copyInto.data));
             }
         }
     }
 
+    if (doneParsingHeader)
+        stopBufferingHeaderBytes();
+
+    if (expectingHeader && !doneParsingHeader) {
+        // more data needed to parse the header
+        copyInto.offset += len;
+        fileRead();
+        return;
+    }
+
     callback(len);
```

Comment:
I agree that the code looks a little strange and probably should be refactored. However, it works correctly because the user basically ignores copyInto.data (and len) when dealing with headers (but it cares about copyInto.offset, where we accumulate the header length).